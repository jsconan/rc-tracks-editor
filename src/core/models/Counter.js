/**
 * RC Tracks Editor
 * Copyright (c) 2022 Jean-SÃ©bastien CONAN
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */

import { validateCallback } from '../helpers';
import { eventEmitterMixin } from '../mixins';

/**
 * Represents a set of counters.
 */
export class Counter extends Map {
    /**
     * Creates an observable set of counters.
     * Each change in the counters will be notified to the subscribers.
     * @param {*} source - An iterable object that can be used to initialize the counters.
     */
    constructor(source = null) {
        super();

        eventEmitterMixin(this);

        if (source) {
            this.load(source);
            source = void 0;
        }
    }

    /**
     * Applies a callback to each counter from the list.
     * @param {function} walker - A callback that will be applied to each counter of the list.
     * @returns {Counter} - Chains the instance.
     * @throws {TypeError} - If the given callback is not a function.
     */
    forEach(walker) {
        validateCallback(walker);

        super.forEach((value, key) => walker.call(this, value, key, this));

        return this;
    }

    /**
     * Maps the counters to an object.
     * @param {function} mapper - A mapper callback that will be applied to each counter.
     * @returns {object}
     * @throws {TypeError} - If the given callback is not a function.
     */
    map(mapper) {
        validateCallback(mapper);

        const counters = {};

        for (const [key, value] of this.entries()) {
            counters[key] = mapper.call(this, value, key, this);
        }

        return counters;
    }

    /**
     * Gets the value for a particular counter.
     * @param {string} key - The key of the counter to read.
     * @returns {number} - The current value of the counter. It returns 0 if the counter does not exist.
     */
    get(key) {
        return super.get(key) || 0;
    }

    /**
     * Sets the value for a particular counter.
     * @param {string} key - The key of the counter to set.
     * @param {number} value  - The new value of the counter. It must be a whole number.
     * @returns {Counter} - Chains the instance.
     * @fires set
     */
    set(key, value) {
        const previous = super.get(key) || 0;
        super.set(key, Math.floor(value) || 0);

        this.emit('set', key, value, previous);

        return this;
    }

    /**
     * Removes a counter.
     * @param {string} key - The key of the counter to delete.
     * @returns {boolean} - Returns `true` if a counter was deleted.
     * @fires delete
     */
    delete(key) {
        const value = super.get(key) || 0;
        const deleted = super.delete(key);

        if (deleted) {
            this.emit('delete', key, value);
        }

        return deleted;
    }

    /**
     * Increments a counter.
     * @param {string} key - The key of the counter to increment.
     * @param {number} amount - The amount to add to the counter.
     * @returns {Counter} - Chains the instance.
     * @fires set
     */
    increment(key, amount = 1) {
        const value = Math.floor(amount) || 1;
        return this.set(key, this.get(key) + value);
    }

    /**
     * Decrements a counter.
     * @param {string} key - The key of the counter to decrement.
     * @param {number} amount - The amount to subtract from the counter.
     * @returns {Counter} - Chains the instance.
     * @fires set
     */
    decrement(key, amount = 1) {
        const value = Math.floor(amount) || 1;
        return this.set(key, this.get(key) - value);
    }

    /**
     * Removes all counters.
     * @returns {Counter} - Chains the instance.
     * @fires clear
     */
    clear() {
        super.clear();

        this.emit('clear');

        return this;
    }

    /**
     * Loads counter from another source. The counters are cleared before.
     * @param {*} iterator - An iterable object that can be used to set the counters.
     * @returns {Counter} - Chains the instance.
     * @fires load
     */
    load(iterator) {
        if (!iterator || !iterator[Symbol.iterator]) {
            return this;
        }

        super.clear();
        for (const [key, value] of iterator) {
            super.set(key, Math.floor(value) || 0);
        }

        this.emit('load');

        return this;
    }

    /**
     * Exports the counters to an object.
     * @returns {object} - An object containing the counters.
     */
    toObject() {
        const counters = {};

        for (const [key, value] of this.entries()) {
            counters[key] = value;
        }

        return counters;
    }

    /**
     * Validates that the given object is an instance of the class.
     * Otherwise, an error is thrown.
     * @param {object} object - The instance to validate.
     * @throws {TypeError} - If the given object is not a valid instance.
     */
    static validateInstance(object) {
        if (!(object instanceof this)) {
            throw new TypeError(`The object must be an instance of ${this.name}!`);
        }
    }
}

/**
 * Notifies a value has been set to a counter.
 * @event set
 * @param {string} key - The key of the counter that was set.
 * @param {number} newValue - The new value.
 * @param {number} oldValue - The previous value.
 */

/**
 * Notifies a counter has been removed.
 * @event delete
 * @param {string} key - The key of the counter that was removed.
 * @param {number} value - The last value of the removed counter.
 */

/**
 * Notifies all counters has been removed.
 * @event clear
 */

/**
 * Notifies the counters has been loaded.
 * @event load
 */
