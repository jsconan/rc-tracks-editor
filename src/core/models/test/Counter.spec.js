/**
 * RC Tracks Editor
 * Copyright (c) 2022 Jean-SÃ©bastien CONAN
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */

import { Counter } from '../Counter.js';

const source = [
    ['c1', 0],
    ['c2', 1],
    ['c3', -1]
];
const keys = ['c1', 'c2', 'c3'];

describe('Counter', () => {
    it('is a class', () => {
        expect(Counter).toEqual(expect.any(Function));
    });

    it('can be initialized with a source', () => {
        expect([...new Counter()]).toEqual([]);
        expect([...new Counter(source)]).toEqual(source);
    });

    it('has a size', () => {
        expect(new Counter().size).toBe(0);
        expect(new Counter(source).size).toBe(3);
    });

    it('implements the iteration protocol', () => {
        const counters = new Counter(source);

        expect(counters[Symbol.iterator]).toEqual(expect.any(Function));
        expect(counters[Symbol.iterator]()).not.toBe(counters[Symbol.iterator]());
        expect([...counters]).toStrictEqual(source);
    });

    it('can produce an iterator for getting the keys', () => {
        const counters = new Counter(source);

        expect(counters.keys).toEqual(expect.any(Function));
        expect(counters.keys()[Symbol.iterator]).toEqual(expect.any(Function));

        expect(counters.keys()).not.toBe(counters.keys());
        expect([...counters.keys()]).toStrictEqual(keys);
    });

    it('can walk over its counters', () => {
        const counters = new Counter(source);

        expect(counters.forEach).toEqual(expect.any(Function));

        const iterator = source[Symbol.iterator]();
        const callback = jest.fn().mockImplementation(function (value, key, thisCounters) {
            expect(this).toBe(counters);
            expect(thisCounters).toBe(counters);
            const [iteratorKey, iteratorValue] = iterator.next().value;
            expect(value).toBe(iteratorValue);
            expect(key).toBe(iteratorKey);
        });

        expect(counters.forEach(callback)).toBe(counters);
        expect(callback).toHaveBeenCalledTimes(source.length);
    });

    it('needs a valid callback to walk over counters', () => {
        expect(() => new Counter().forEach()).toThrow('A callback function is expected!');
    });

    it('can map its counters', () => {
        const counters = new Counter(source);

        expect(counters.map).toEqual(expect.any(Function));

        const iterator = source[Symbol.iterator]();
        const callback = jest.fn().mockImplementation(function (value, key, thisCounter) {
            expect(this).toBe(counters);
            expect(thisCounter).toBe(counters);
            const [iteratorKey, iteratorValue] = iterator.next().value;
            expect(value).toBe(iteratorValue);
            expect(key).toBe(iteratorKey);
            return value * 10;
        });

        const mapped = counters.map(callback);
        expect(mapped).not.toBe(counters);
        expect(mapped).toMatchSnapshot();
        expect(callback).toHaveBeenCalledTimes(source.length);
    });

    it('needs a valid callback to map counters', () => {
        expect(() => new Counter().map()).toThrow('A callback function is expected!');
    });

    it('can tell if a counter exists', () => {
        const counters = new Counter(source);

        expect(counters.has('c1')).toBeTruthy();
        expect(counters.has('c2')).toBeTruthy();
        expect(counters.has('c3')).toBeTruthy();
        expect(counters.has('c4')).toBeFalsy();
    });

    it('can get the value of a counter', () => {
        const counters = new Counter(source);

        expect(counters.get('c1')).toBe(0);
        expect(counters.get('c2')).toBe(1);
        expect(counters.get('c3')).toBe(-1);
        expect(counters.get('c4')).toBe(0);
    });

    it('can set the value of a counter', () => {
        const counters = new Counter(source);

        expect(counters.get('c1')).toBe(0);
        expect(counters.set('c1', 10)).toBe(counters);
        expect(counters.get('c1')).toBe(10);

        expect(counters.get('c4')).toBe(0);
        expect(counters.set('c4', 1)).toBe(counters);
        expect(counters.get('c4')).toBe(1);
    });

    it('can only set whole numbers', () => {
        const counters = new Counter(source);

        expect(counters.get('c1')).toBe(0);
        expect(counters.set('c1', 1.2)).toBe(counters);
        expect(counters.get('c1')).toBe(1);

        expect(counters.set('c1', 2.9)).toBe(counters);
        expect(counters.get('c1')).toBe(2);

        expect(counters.set('c1', 'c')).toBe(counters);
        expect(counters.get('c1')).toBe(0);
    });

    it('can delete a counter', () => {
        const counters = new Counter(source);

        expect(counters.has('c1')).toBeTruthy();
        expect(counters.delete('c1')).toBeTruthy();
        expect(counters.has('c1')).toBeFalsy();
        expect(counters.get('c1')).toBe(0);
        expect(counters.delete('c1')).toBeFalsy();
    });

    it('can increment a counter', () => {
        const counters = new Counter();

        expect(counters.get('c1')).toBe(0);
        expect(counters.increment('c1')).toBe(counters);
        expect(counters.get('c1')).toBe(1);
        expect(counters.increment('c1', 2)).toBe(counters);
        expect(counters.get('c1')).toBe(3);

        expect(counters.increment('c1', 2.5)).toBe(counters);
        expect(counters.get('c1')).toBe(5);

        expect(counters.increment('c1', 0)).toBe(counters);
        expect(counters.get('c1')).toBe(6);
    });

    it('can decrement a counter', () => {
        const counters = new Counter();

        expect(counters.get('c1')).toBe(0);
        expect(counters.decrement('c1')).toBe(counters);
        expect(counters.get('c1')).toBe(-1);
        expect(counters.decrement('c1', 2)).toBe(counters);
        expect(counters.get('c1')).toBe(-3);

        expect(counters.decrement('c1', 2.5)).toBe(counters);
        expect(counters.get('c1')).toBe(-5);

        expect(counters.decrement('c1', 0)).toBe(counters);
        expect(counters.get('c1')).toBe(-6);
    });

    it('can removes all counters', () => {
        const counters = new Counter(source);

        expect([...counters]).toStrictEqual(source);
        expect(counters.clear()).toBe(counters);
        expect([...counters]).toStrictEqual([]);
    });

    it('can load counters', () => {
        const counters = new Counter();
        counters.set('c', 1);

        expect([...counters]).toStrictEqual([['c', 1]]);
        expect(counters.load({})).toBe(counters);
        expect([...counters]).toStrictEqual([['c', 1]]);

        expect(counters.load(source)).toBe(counters);
        expect([...counters]).toStrictEqual(source);
    });

    it('can reset counters', () => {
        const counters = new Counter();
        counters.set('a', 2);
        counters.set('b', 3);

        expect([...counters]).toStrictEqual([
            ['a', 2],
            ['b', 3]
        ]);
        expect(counters.reset()).toBe(counters);
        expect([...counters]).toStrictEqual([
            ['a', 0],
            ['b', 0]
        ]);

        expect(counters.reset(['a', 'b', 'c', 'a', 'c'])).toBe(counters);
        expect([...counters]).toStrictEqual([
            ['a', 2],
            ['b', 1],
            ['c', 2]
        ]);
    });

    it('can export counters', () => {
        const counters = new Counter(source);

        expect(counters.toObject()).toMatchSnapshot();
    });

    it('can notify changes', () => {
        const counters = new Counter(source);

        const callback = jest.fn().mockImplementation(value => {
            expect(value).toBe(counters);
            expect([...value]).toMatchSnapshot();
        });

        const unsubscribe = counters.subscribe(callback); // callback called

        counters.notify(); // callback called
        counters.set('c1', 3); // callback called
        counters.delete('c2'); // callback called
        counters.delete('c2');
        counters.increment('c2'); // callback called
        counters.decrement('c2'); // callback called
        counters.clear(); // callback called
        counters.load(source); // callback called
        counters.load({});
        counters.reset(source); // callback called
        counters.reset({}); // callback called

        unsubscribe();
        counters.increment('c1');

        expect(callback).toHaveBeenCalledTimes(10);
    });
});
