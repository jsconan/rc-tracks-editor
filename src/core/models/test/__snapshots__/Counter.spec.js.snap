// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Counter can export counters 1`] = `
Object {
  "c1": 0,
  "c2": 1,
  "c3": -1,
}
`;

exports[`Counter can map its counters 1`] = `
Object {
  "c1": 0,
  "c2": 10,
  "c3": -10,
}
`;

exports[`Counter can notify changes 1`] = `
Array [
  Array [
    "c1",
    0,
  ],
  Array [
    "c2",
    1,
  ],
  Array [
    "c3",
    -1,
  ],
]
`;

exports[`Counter can notify changes 2`] = `
Array [
  Array [
    "c1",
    0,
  ],
  Array [
    "c2",
    1,
  ],
  Array [
    "c3",
    -1,
  ],
]
`;

exports[`Counter can notify changes 3`] = `
Array [
  Array [
    "c1",
    3,
  ],
  Array [
    "c2",
    1,
  ],
  Array [
    "c3",
    -1,
  ],
]
`;

exports[`Counter can notify changes 4`] = `
Array [
  Array [
    "c1",
    3,
  ],
  Array [
    "c3",
    -1,
  ],
]
`;

exports[`Counter can notify changes 5`] = `
Array [
  Array [
    "c1",
    3,
  ],
  Array [
    "c3",
    -1,
  ],
  Array [
    "c2",
    1,
  ],
]
`;

exports[`Counter can notify changes 6`] = `
Array [
  Array [
    "c1",
    3,
  ],
  Array [
    "c3",
    -1,
  ],
  Array [
    "c2",
    0,
  ],
]
`;

exports[`Counter can notify changes 7`] = `Array []`;

exports[`Counter can notify changes 8`] = `
Array [
  Array [
    "c1",
    0,
  ],
  Array [
    "c2",
    1,
  ],
  Array [
    "c3",
    -1,
  ],
]
`;

exports[`Counter can notify changes 9`] = `
Array [
  Array [
    "c1",
    0,
  ],
  Array [
    "c2",
    0,
  ],
  Array [
    "c3",
    0,
  ],
  Array [
    Array [
      "c1",
      0,
    ],
    1,
  ],
  Array [
    Array [
      "c2",
      1,
    ],
    1,
  ],
  Array [
    Array [
      "c3",
      -1,
    ],
    1,
  ],
]
`;

exports[`Counter can notify changes 10`] = `
Array [
  Array [
    "c1",
    0,
  ],
  Array [
    "c2",
    0,
  ],
  Array [
    "c3",
    0,
  ],
  Array [
    Array [
      "c1",
      0,
    ],
    0,
  ],
  Array [
    Array [
      "c2",
      1,
    ],
    0,
  ],
  Array [
    Array [
      "c3",
      -1,
    ],
    0,
  ],
]
`;
